src/ripple/protocol/SField.h

```diff
@@ -51,44 +51,67 @@ template <class>
 class STInteger;
 class STXChainBridge;
 class STVector256;
-
-enum SerializedTypeID {
-    // special types
-    STI_UNKNOWN = -2,
-    STI_NOTPRESENT = 0,
-
-    // // types (common)
-    STI_UINT16 = 1,
-    STI_UINT32 = 2,
-    STI_UINT64 = 3,
-    STI_UINT128 = 4,
-    STI_UINT256 = 5,
-    STI_AMOUNT = 6,
-    STI_VL = 7,
-    STI_ACCOUNT = 8,
-    // 9-13 are reserved
-    STI_OBJECT = 14,
-    STI_ARRAY = 15,
-
-    // types (uncommon)
-    STI_UINT8 = 16,
-    STI_UINT160 = 17,
-    STI_PATHSET = 18,
-    STI_VECTOR256 = 19,
-    STI_UINT96 = 20,
-    STI_UINT192 = 21,
-    STI_UINT384 = 22,
-    STI_UINT512 = 23,
-    STI_ISSUE = 24,
-    STI_XCHAIN_BRIDGE = 25,
-
-    // high level types
-    // cannot be serialized inside other types
-    STI_TRANSACTION = 10001,
-    STI_LEDGERENTRY = 10002,
-    STI_VALIDATION = 10003,
-    STI_METADATA = 10004,
-};
+class Definitions;
+
+#pragma push_macro("XMACRO")
+#undef XMACRO
+
+#define XMACRO(STYPE)                             \
```

Comment:
The advantage is that all of the contents of `SerializedTypeID` are in a map. There's no way to iterate through an enum in C++. The macro ensures that you only need to have one line of code for each type ID.

Reply:
i see okay. 

Do we need both `SerializedTypeID` and `sTypeMap`? 
Could we get rid of `enum SerializedTypeID` and live with just the map? (because their contents are identical) I feel that would obviate the macros

Reply:
No, the enum is used all over the repo, and it would likely slow down a lot of rippled to force it to call a map.
The contents are supposed to be identical - it needs to be both in an enum _and_ in a map. The enum is for ease of use all over the codebase, and the map is specifically for `server_definitions`, so you can iterate through the map instead of needing to somehow iterate through the enum.

Reply:
I see, alright. 

is this because of performance difference between accessing an `enum` versus `std::unordered_map` or is it more of developer-ease?

Reply:
Yes.